### PR TITLE
Implement TICK_SIZE for bitmex

### PR DESCRIPTION
**Pull request by brandsimon**
Good day everyone,

as discussed in #5055 there is a problem in the bitmex precision handling code.
Bitmex uses tickSize to tell us, what precision is allowed.
e.g. the price for BTCUSD has a tickSize of 0.5.
The current code allows a precision of 1 here, this means, 0.1 would also be correct, which it is not.

This code is hopefully fixing this issue. Please let me know if there are changes needed.

### CODE REVIEW

**Comment by kroitor** [comment]
@brandsimon it's great to hear from you again! ) I'll try the private tests on my side and if no edge-cases pop up – will merge it very quickly, hopefully, this weekend ) Thank you! )

**Comment by kroitor** [comment]
Should we enforce it now here https://github.com/ccxt/ccxt/pull/5256/files#diff-c23b6c3f281ff38df587025321e40445R1119 as we do, say, here: https://github.com/ccxt/ccxt/blob/master/js/hitbtc2.js#L1030 ?

**Comment by brandsimon** [comment]
@kroitor Thank you very much! :)
I would rather remove it from the hitbtc2 implementation.
Is there a reason this is enforced only on some exchanges and not on all?
I think mixed behavior here only leads to bugs in some edge cases.
If the software developer wants to send a wrong amount, he does not see/know, that the order size changed unless he checks the order. (I'd rather see an exception from the exchange)

**Comment by kroitor** [comment]
> Is there a reason this is enforced only on some exchanges and not on all?

Yes – this was planned to be enforced on all exchanges, but on BitMEX we could not do that because we didn't have the support for TICK_SIZE.

> I think mixed behavior here only leads to bugs in some edge cases.

That's true, however, we suggest to add it everywhere, rather than remove it from everywhere.
The reasoning behind that is simple: when the user submits an amount of 0.123456789 to an exchange with the `DECIMAL_PLACES` precision mode and `precision['amount'] === 4`, the amount is truncated to 0.1234 and sent this way to the exchange. The price is rounded to `precision['price']`. This logic is financially correct. On the other hand, if would be very hard for the new users to even understand what the system wants from them if they had to do the precision in userland. So, this is implemented everywhere but BitMEX, and I'd rather add it to BitMEX instead of cutting it away )

In other words, we use `amountToPrecision` and `priceToPrecision` everywhere (which are wrappers for `decimalToPrecision`).

Are you ok if we add it to BitMEX to keep it consistent with other implementations? Lmk your thoughts or if anything )

**Comment by brandsimon** [comment]
@kroitor 
Thank you for the detailed explanation!
If it should be enforced for every exchange, I am totally ok with adding this to bitmex too.

I dont understand why you assume the user would want `0.1234` instead of `0.1235`. Why should the user always truncate amounts?
Example: A user wants to test something with a really low amount, so he chooses `0.00001`, but this gets truncated to `0.0` which results in an exchange exception. The user will be really confused if he knows nothing about this. (The Law of Leaky Abstractions) Without this, he would get an excpetion, that the amount is not supported by the exchange and hopefully find the needed functions/wiki pages per search engine.

I agree, if this concept is abstracted away from the user, it will be easier for new users to start using it, but only until they encounter behavior they dont expect.
Personally I see ccxt as a low-level API and I dont expect low-level APIs to abstract things away. I see such an behavior in a higher level API.

For myself I can still do the precision handling in userland, the function wont change the amount/price afterwards, so this behavior is ok with me.

**Comment by kroitor** [comment]
> I dont understand why you assume the user would want 0.1234 instead of 0.1235. Why should the user always truncate amounts?

Oh, let me explain the reasoning behind it. I'm not saying that the user would **always** want 0.1234 instead of 0.1235.

First and foremost, we'd want to keep the lib "plug-and-play". That implies that a call to `createOrder` should generally work without too much of a hassle. If you omit all the hassle, the essence that remains is:

```Python
# Python here just for simplicity
import ccxt
exchange = ccxt.binance ({ 'apiKey': KEY, 'secret': SECRET })
exchange.create_order('BTC/USDT', 'limit', 'buy', 1, 8000)
```

↑ We'd like to keep the lib as easy as the above snippet. For that to work, we have to abstract away the exchanges' precision and limits.

Now, ignore CCXT for a second and suppose the user wants to buy an amount of 0.12345 and he is talking to an exchange directly and the exchange has precision 4. In that case the user can send 0.12345 and will get an error from the exchange. After that the user has a choice: either send 0.1234 or 0.1235. If the user wanted his amount to be rounded to 0.1235, he would specify the value of 0.1235 in the first place. 

The exchanges' rules for precision can differ a lot, for example Bitfinex required a float price to precision, where other exchanges require a string-float. And if we don't abstract that away, the code to handle the order placement will have to be bloated to something like:

```Python
# Python here just for simplicity
import ccxt
exchange = ccxt.binance ({ 'apiKey': KEY, 'secret': SECRET })
amount = 0.1234
if exchange.precisionMode == 'A':
    amount = truncate
if exchnage.precisionMode == 'B':
    amount = truncate

# + tons of other exchange-specific checks

exchange.create_order('BTC/USDT', 'limit', 'buy', 1, 8000)
```

↑ It becomes much harder to even start using this lib then, the user would have to add quite a lot of exchange-specific checks, and we'd like to avoid having exchange-specific code outside of the lib as much as possible, for the purposes of unification. In the end, even after adding all that unncessary bloated code, when placing an order the user would still want to "spend money up to the designated amount", but not above it.

So, the reason for truncation is that we can't avoid abstracting some things, and if we abstract some of them, then truncation seems the most resonable. Definitely not rounding.

Also, if we account for the order filling process, the order in general will not get filled up completely, leaving some dust – that's common with practically all exchanges and that's another reason for truncation (as long as we have to abstract it somehow anyway). In other words, truncation is something that happens within the exchanges' engines themselves for practically any of your orders.

> Example: A user wants to test something with a really low amount, so he chooses 0.00001, but this gets truncated to 0.0 which results in an exchange exception.

The user will get an exchange error in both cases, so, the "zero"-scenario is easy to handle )

Anyways, I am merging this, and will add the edits to `createOrder` in `bitmex`. Thank you again for your help and your involvement! )

**Comment by kroitor** [comment]
@brandsimon looks like we have an issue with TICK_SIZE not formatting some numbers properly:

- https://github.com/ccxt/ccxt/issues/6483

```
assert (decimalToPrecision ('0.0006', TRUNCATE, 0.0001, TICK_SIZE) === '0.0006');
```

↑ This yields 0.0005 instead of the expected result of 0.0006.

I'd be very thankful if you could assist in resolving this. Your involvement is appreciated a lot!

**Comment by brandsimon** [comment]
@kroitor 
Thank you for the notification. I will have a look. (probably tomorrow).
